internal/controller/instaslice_controller.go

```diff
@@ -114,55 +114,104 @@ func (r *InstasliceReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	// failed pods are not deleted by InstaSlice, finalizer is removed so that user can
 	// delete the pod and later InstaSlice will clean up the MIG slice.
 	if pod.Status.Phase == v1.PodFailed && controllerutil.ContainsFinalizer(pod, finalizerOrGateName) {
-
-		errRemovingFinalizer := controllerutil.RemoveFinalizer(pod, finalizerOrGateName)
-		if !errRemovingFinalizer {
-			log.FromContext(ctx).Info("finalizer not deleted for ", "pod", pod.Name)
+		allocationFound := false
+		for _, instaslice := range instasliceList.Items {
+			for podUuid, allocation := range instaslice.Spec.Allocations {
+				if podUuid == string(pod.UID) {
+					allocationFound = true
+					if allocation.Allocationstatus == inferencev1alpha1.AllocationStatusDeleted {
+						resultDelete, errDeletingAllocation := r.deleteInstasliceAllocation(ctx, instaslice, allocation)
+						if err != nil {
+							return resultDelete, errDeletingAllocation
+						}
+						resultRemove, errRemovingFinalizer := r.removeInstaSliceFinalizer(ctx, req)
+						if err != nil {
+							return resultRemove, errRemovingFinalizer
+						}
+					} else {
+						log.FromContext(ctx).Info("pod failed set allocation status to deleting ", "pod", allocation.PodName)
+						allocation.Allocationstatus = inferencev1alpha1.AllocationStatusDeleting
+						var updateInstasliceObject inferencev1alpha1.Instaslice
+						typeNamespacedName := types.NamespacedName{
+							Name:      instaslice.Name,
+							Namespace: "default", // TODO: modify
+						}
+						err := r.Get(ctx, typeNamespacedName, &updateInstasliceObject)
+						if err != nil {
+							log.FromContext(ctx).Error(err, "error getting latest instaslice object")
+						}
+						updateInstasliceObject.Spec.Allocations[podUuid] = allocation
+						errUpdatingInstaslice := r.Update(ctx, &updateInstasliceObject)
+						if errUpdatingInstaslice != nil {
+							log.FromContext(ctx).Info("unable to set instaslice to state deleted for ", "pod", allocation.PodName)
+							return ctrl.Result{RequeueAfter: 1 * time.Second}, nil
+						}
+					}
```

Comment:
The change seems good but most of this code i.e. highlighted code snippet seems to be repetitive.
One suggestion could be
1. Get the Instaslice object
2. Based on the pod condition, update the desired allocations etc..
3. Remove the Finalizer
4. Update the Instaslice with the desired values that are set earlier
This way, we can reduce the repeated code and hence the code redundancy
What do you think?